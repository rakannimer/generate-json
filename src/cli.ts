const mri = require("mri");
const fs = require("fs");
const mkdirp = require("mkdirp");
const { generateJson } = require("./index");
const pkg = require("../package.json");
console.log(`
Version : ${pkg.version}
`);

const args = process.argv.slice(2);

const { _ } = mri(args);

if (_.length < 2) {
  throw new Error(`Expected 2 arguments, <schemaFilePath> <outputFilePath>`);
}
const [schemaFilePath, outputDirPath] = _;
const schemaFile = require(`${process.cwd()}/${schemaFilePath}`);
const { schema, keyReducers, count } = schemaFile;

const { tree, values, keys } = generateJson({ schema, keyReducers, count });
const outputDir = `${process.cwd()}/${outputDirPath}`;
mkdirp(outputDir);
fs.writeFileSync(`${outputDir}/tree.json`, JSON.stringify(tree), {
  encoding: "utf8"
});
fs.writeFileSync(`${outputDir}/values.json`, JSON.stringify(values), {
  encoding: "utf8"
});
fs.writeFileSync(`${outputDir}/keys.json`, JSON.stringify(keys), {
  encoding: "utf8"
});
console.log(`âœ… Ok. 
Check ${outputDir} 
to see the generated data.
`);
