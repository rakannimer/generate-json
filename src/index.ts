import flatten from "flat";
import faker from "faker";
import set from "lodash.set";

export type StringNode = {
  depth: number;
  value: string;
};

export const generateKey = (
  keyName: string,
  depth: number,
  keyReducers = {} as any,
  keyMap: Map<string, StringNode[]>
) => {
  let customGenerateKey = (fakerjs: typeof faker) =>
    `G_K_${fakerjs.random.alphaNumeric(5)}`;
  if (keyName in keyReducers) {
    customGenerateKey = keyReducers[keyName];
  }

  const cacheKey = `${keyName}`;
  let keyValue;
  if (!keyMap.has(cacheKey)) {
    keyValue = customGenerateKey(faker);
  } else {
    const previouslyGeneratedKeys = keyMap.get(cacheKey) as StringNode[];
    const ancestorKeys = previouslyGeneratedKeys.filter(n => n.depth <= depth);
    if (ancestorKeys.length > 0) {
      keyValue = faker.random.arrayElement(previouslyGeneratedKeys).value;
    } else {
      keyValue = customGenerateKey(faker);
    }
  }
  keyMap.set(cacheKey, [{ value: keyValue, depth }]);
  return keyValue;
};

export const generateData = (schema: any, keyReducers = {}) => {
  const keyMap = new Map() as Map<string, StringNode[]>;

  const flatSchema = flatten(schema);
  const flatSchemaKeys = Object.keys(flatSchema);
  const flatSchemaValues = Object.values(flatSchema);
  const depth = flatSchemaKeys.reduce(
    (acc, cur) => Math.max(acc, cur.split(".").length),
    0
  );
  const isVar = (s: string) => {
    return s.startsWith("{") && s.endsWith("}");
  };
  const keys = flatSchemaKeys.map(key => {
    return key
      .split(".")
      .map((s, i) => ({ value: s, depth: i }))
      .map(val => ({ isVar: isVar(val.value), ...val }))
      .map(varInPath => {
        const { isVar, value, depth } = varInPath;
        if (isVar) {
          return generateKey(value, depth, keyReducers, keyMap);
        } else {
          return value;
        }
      })
      .join(".");
  });
  const values = flatSchemaValues.map((valTypeOrReference, i) => {
    let generatedValue;
    if (typeof valTypeOrReference === "function") {
      generatedValue = valTypeOrReference(faker, keys[i]);
    } else if (isVar(valTypeOrReference)) {
      generatedValue = generateKey(valTypeOrReference, depth + 1, {}, keyMap);
    } else {
      switch (valTypeOrReference) {
        case "number": {
          generatedValue = faker.random.number();
          break;
        }
        case "string": {
          generatedValue = faker.random.words(2);
          break;
        }
        case "boolean": {
          generatedValue = faker.random.boolean();
          break;
        }
        case "timestamp": {
          generatedValue = faker.date.future(1, new Date(2018)).getTime();
          break;
        }
        default: {
          // console.log(`UNKNOWN ${valTypeOrReference}`);
          generatedValue = valTypeOrReference;
        }
      }
    }
    return generatedValue;
  });
  const tree = keys.reduce((acc, cur, i) => {
    set(acc, cur, values[i]);
    return acc;
  }, {});
  return { keys, values, tree };
};

export const generateJsonDefaultArgs = {
  schema: {} as any,
  keyReducers: {} as any,
  count: 1 as any
} as GenerateJsonArgs;

export type GenerateJsonArgs = {
  schema: any;
  keyReducers?: any;
  count?: number;
};

export const generateJson = ({
  schema = {},
  keyReducers = {},
  count = 1
} = generateJsonDefaultArgs) => {
  let jsonKeys = [] as string[];
  let jsonValues = [] as any[];
  let jsonTree = {};
  for (let i = 0; i < count; i += 1) {
    const { keys, values, tree } = generateData(schema, keyReducers);
    jsonKeys = [...keys, ...jsonKeys];
    jsonValues = [...values, ...jsonValues];
    jsonTree = {
      ...jsonTree,
      ...tree
    };
  }
  return { keys: jsonKeys, values: jsonValues, tree: jsonTree };
};
