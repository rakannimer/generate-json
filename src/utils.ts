import admin from "firebase-admin";
import is from "is";
export const isPromise = (promiseMaybe: any) => {
  return (
    !!promiseMaybe &&
    (typeof promiseMaybe === "object" || typeof promiseMaybe === "function") &&
    typeof promiseMaybe.then === "function"
  );
};

export const getValueAt = async (path: string) => {
  const val = await admin
    .app()
    .database()
    .ref(path)
    .once("value");
  return val.val();
};

export const getType = (val: any) => {
  if (is.boolean(val)) {
    return "boolean";
  } else if (is.number(val)) {
    return "number";
  } else if (is.string(val)) {
    return "string";
  } else if (is.null(val)) {
    return "null";
  } else if (is.undefined(val)) {
    return "undefined";
  } else {
    // console.log(val);
    throw val;
  }
};
