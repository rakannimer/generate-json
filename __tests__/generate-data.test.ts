// import { generateData } from "../dist/index.cjs";
import { generateData } from "../src/";

import faker from "faker";

const SEED = 42;

const keyReducers = [
  {
    "{env}": (fakerjs: typeof faker) => {
      return fakerjs.random.arrayElement(["prod", "staging"]);
    },
    "{userID}": (fakerjs: typeof faker) => {
      return fakerjs.random.alphaNumeric(5);
    },
    "{postID}": (fakerjs: typeof faker) => {
      return fakerjs.random.alphaNumeric(5);
    }
  }
];

const schemas = [
  {
    "{env}": {
      users: {
        "{userID}": {
          comment_count: "number",
          username: "string",
          is_verified: "boolean",
          created_at: "timestamp"
        }
      },
      posts: {
        "{postID}": {
          post_title: "string",
          post_data: (faker, key) => {
            // return faker.random.words(5)
            return key.split("/").join("-") + ":post_data";
          },
          author_id: "{userID}",
          created_at: "timestamp"
        }
      },
      comments: {
        "{postID}": {
          "{commentID}": {
            value: "üëç",
            userID: "{userID}"
          }
        }
      }
    }
  }
];

for (let i = 0; i < schemas.length; i += 1) {
  const schema = schemas[0];
  const keyReducer = keyReducers[i];
  test("generateData works without second argument", () => {
    faker.seed(SEED);
    const data = generateData(schema);
    expect(data).toMatchSnapshot();
  });

  test("generateData works with second argument", () => {
    faker.seed(SEED);
    const data = generateData(schema, keyReducer);
    // console.log(data);
    expect(data).toMatchSnapshot();
  });
}
