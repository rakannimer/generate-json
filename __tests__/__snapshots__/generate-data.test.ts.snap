// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateData works with second argument 1`] = `
Object {
  "keys": Array [
    "prod.users.sy6qs.comment_count",
    "prod.users.sy6qs.username",
    "prod.users.sy6qs.is_verified",
    "prod.users.sy6qs.created_at",
    "prod.posts.gvcl5.post_title",
    "prod.posts.gvcl5.post_data",
    "prod.posts.gvcl5.author_id",
    "prod.posts.gvcl5.created_at",
    "prod.comments.gvcl5.G_K_706z6.value",
    "prod.comments.gvcl5.G_K_706z6.userID",
  ],
  "tree": Object {
    "prod": Object {
      "comments": Object {
        "gvcl5": Object {
          "G_K_706z6": Object {
            "userID": "sy6qs",
            "value": "üëç",
          },
        },
      },
      "posts": Object {
        "gvcl5": Object {
          "author_id": "sy6qs",
          "created_at": 14382625758,
          "post_data": "prod.posts.gvcl5.post_data:post_data",
          "post_title": "enable Gorgeous Frozen Pants",
        },
      },
      "users": Object {
        "sy6qs": Object {
          "comment_count": 52475,
          "created_at": 19295395373,
          "is_verified": true,
          "username": "Grocery Home",
        },
      },
    },
  },
  "values": Array [
    52475,
    "Grocery Home",
    true,
    19295395373,
    "enable Gorgeous Frozen Pants",
    "prod.posts.gvcl5.post_data:post_data",
    "sy6qs",
    14382625758,
    "üëç",
    "sy6qs",
  ],
}
`;

exports[`generateData works without second argument 1`] = `
Object {
  "keys": Array [
    "G_K_dsy6q.users.G_K_sll5g.comment_count",
    "G_K_dsy6q.users.G_K_sll5g.username",
    "G_K_dsy6q.users.G_K_sll5g.is_verified",
    "G_K_dsy6q.users.G_K_sll5g.created_at",
    "G_K_dsy6q.posts.G_K_5pn02.post_title",
    "G_K_dsy6q.posts.G_K_5pn02.post_data",
    "G_K_dsy6q.posts.G_K_5pn02.author_id",
    "G_K_dsy6q.posts.G_K_5pn02.created_at",
    "G_K_dsy6q.comments.G_K_5pn02.G_K_6mami.value",
    "G_K_dsy6q.comments.G_K_5pn02.G_K_6mami.userID",
  ],
  "tree": Object {
    "G_K_dsy6q": Object {
      "comments": Object {
        "G_K_5pn02": Object {
          "G_K_6mami": Object {
            "userID": "G_K_sll5g",
            "value": "üëç",
          },
        },
      },
      "posts": Object {
        "G_K_5pn02": Object {
          "author_id": "G_K_sll5g",
          "created_at": 24761311291,
          "post_data": "G_K_dsy6q.posts.G_K_5pn02.post_data:post_data",
          "post_title": "Coordinator Multi-channelled",
        },
      },
      "users": Object {
        "G_K_sll5g": Object {
          "comment_count": 29122,
          "created_at": 9213076313,
          "is_verified": false,
          "username": "Namibia enable",
        },
      },
    },
  },
  "values": Array [
    29122,
    "Namibia enable",
    false,
    9213076313,
    "Coordinator Multi-channelled",
    "G_K_dsy6q.posts.G_K_5pn02.post_data:post_data",
    "G_K_sll5g",
    24761311291,
    "üëç",
    "G_K_sll5g",
  ],
}
`;
